### PR TITLE
Thresh count & bivariate spell in sdba properties

### PR DESCRIPTION
**Pull request by coxipi**
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [x] Tests for the changes have been added (for bug fixes / features)
  - [ ] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] CHANGES.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* `spell_length_distribution` now accepts an argument `window`. The default is `window=1` and corresponds to the old behaviour
*  `threshold_count` hard codes `window=1`  in  `spell_length_distribution` (so exactly the same as the default for `spell_length_distribution`
* `double_spell_length_distribution` and `double_threshold_count` for analog function where we have two variables & two thresholds
* finer control on aggregating operations: `stat` and `stat_resample` to control the time reduction operation in and out of resampling 

### Does this PR introduce a breaking change?

No

### Other information:

`threshold_count` might be seen as superfluous. There might be some optimizations available in this case, it might be more convienient for users to have this specific option. But I'm open to just remove this options.

### CODE REVIEW

**Comment by coxipi** [comment]
@Zeitsperre I just ran a notebook with this env:
```python
from __future__ import annotations
import numpy as np
import xarray as xr
from xclim.ensembles import fractional_uncertainty, hawkins_sutton, lafferty_sriver
from xclim.ensembles._filters import _concat_hist, _model_in_all_scens, _single_member
from pathlib import Path
from xclim.testing import open_dataset

"""Just a smoke test."""
dims = {"run": "member", "scen": "scenario"}
da = (
    open_dataset("uncertainty_partitioning/cmip5_pr_global_mon.nc")
    .pr.sel(time=slice("1950", None))
    .rename(dims)
)
da1 = _model_in_all_scens(da)
dac = _concat_hist(da1, scenario="historical")
das = _single_member(dac)
hawkins_sutton(das)
```
and I don't get an error, for `xarray==2024.{03,05}`

**Comment by coxipi** [comment]
For context of the PR, I wanted a function that can be used in a xscen workflow to get measures of dailyfreezethaw:
```yaml
  dlyfrzthw:
    base: xclim.sdba.properties.double_threshold_count
    cf_attrs:
      long_name: "Number of days where maximum daily temperatures are above 0 degC and minimum daily temperatures are  at or below 0 degC."
    input:
      da1: tasmax
      da2: tasmin
    parameters: 
      op1: '>'
      op2: '<='
      thresh1: '0 degC'
      thresh2: '0 degC'
      method1: 'amount'
      method2: 'amount'
      group: 'time'
      stat: 'mean'
  ```
I get the idea we want to separate `xclim` and `sdba` as much as possible, but I just find that `xscen` is trying to leverage both, but in the process, we end up rewritting `xclim` code in `sdba`. It seems pretty clear indicators belong to `xclim`, but then again, many functions in xclim.indices.generic could server xclim AND sdba. 

Perhaps there is a way to have more generic classes/functions in `sdba` that can eventually use a function from `xclim` as an input? Would we want to allow this kind of flexibility in `sdba`?

File: xclim/sdba/properties.py [hunk 1]
```diff
@@ -386,6 +405,70 @@ def _spell_stats(ds, *, dim, method, thresh, op, freq, resample_before_rl, stat)
...
+      Str with units if the method is "amount".
+      Float of the quantile if the method is "quantile".
+    stat: {'mean', 'sum', 'max','min'}
+      Statistics to apply to the remaining time dimension after resampling (e.g. Jan 1980-2010)
+    stat_resample: {'mean', 'sum', 'max','min'}, optional
+      Statistics to apply to the resampled input at the {group} (e.g. 1-31 Jan 1980). If `None`, the same method as `stat` will be used.
+    group : {'time', 'time.season', 'time.month'}
+      Grouping of the output.
+      e.g. For 'time.month', the correlation would be calculated on each month separately,
+      but with all the years together.
```
Comment (Zeitsperre) [inline xclim/sdba/properties.py:444]:
```suggestion
    method : {'amount', 'quantile'}
      Method to choose the threshold.
      'amount': The threshold is directly the quantity in {thresh}. It needs to have the same units as {da}.
      'quantile': The threshold is calculated as the quantile {thresh} of the distribution.
    op : {">", "<", ">=", "<="}
      Operation to verify the condition for a spell.
      The condition for a spell is variable {op} threshold.
    thresh : str or float
      Threshold on which to evaluate the condition to have a spell.
      Str with units if the method is "amount".
      Float of the quantile if the method is "quantile".
    stat : {'mean', 'sum', 'max','min'}
      Statistics to apply to the remaining time dimension after resampling (e.g. Jan 1980-2010)
    stat_resample : {'mean', 'sum', 'max','min'}, optional
      Statistics to apply to the resampled input at the {group} (e.g. 1-31 Jan 1980). If `None`, the same method as `stat` will be used.
    group : {'time', 'time.season', 'time.month'}
      Grouping of the output.
      e.g. For 'time.month', the correlation would be calculated on each month separately,
      but with all the years together.
```

File: xclim/sdba/properties.py [hunk 2]
```diff
@@ -386,6 +405,70 @@ def _spell_stats(ds, *, dim, method, thresh, op, freq, resample_before_rl, stat)
...
+
+    Returns
+    -------
+    xr.DataArray, [dimensionless]
+      {stat} number of days when the variable is {op} the {method} {thresh}.
```
Comment (Zeitsperre) [inline xclim/sdba/properties.py:449]:
```suggestion
        {stat} number of days when the variable is {op} the {method} {thresh}.
```

File: xclim/sdba/properties.py [hunk 3]
```diff
@@ -730,6 +813,245 @@ def _first_output(ds, *, dim, index, corr_type):
...
+      E.g. If 'time.month', the spell lengths are computed separately for each month.
+    resample_before_rl : bool
+      Determines if the resampling should take place before or after the run
+      length encoding (or a similar algorithm) is applied to runs.
+
+    Returns
+    -------
+    xr.DataArray, [units of the sampling frequency]
+      {stat} of spell length distribution when the first variable is {op1} the {method1} {thresh1}
+      and the second variable is {op2} the {method2} {thresh2} for {window} consecutive day(s).
```
Comment (Zeitsperre) [inline xclim/sdba/properties.py:883]:
```suggestion
    da1 : xr.DataArray
        First variable on which to calculate the diagnostic.
    da2 : xr.DataArray
        Second variable on which to calculate the diagnostic.
    method1 : {'amount', 'quantile'}
        Method to choose the threshold.
        'amount': The threshold is directly the quantity in {thresh}. It needs to have the same units as {da}.
        'quantile': The threshold is calculated as the quantile {thresh} of the distribution.
    method2 : {'amount', 'quantile'}
        Method to choose the threshold.
        'amount': The threshold is directly the quantity in {thresh}. It needs to have the same units as {da}.
       'quantile': The threshold is calculated as the quantile {thresh} of the distribution.
    op1 : {">", "<", ">=", "<="}
        Operation to verify the condition for a spell.
        The condition for a spell is variable {op1} threshold.
    op2 : {">", "<", ">=", "<="}
        Operation to verify the condition for a spell.
        The condition for a spell is variable {op2} threshold.
    thresh1 : str or float
        Threshold on which to evaluate the condition to have a spell.
        String with units if the method is "amount".
        Float of the quantile if the method is "quantile".
    thresh2 : str or float
        Threshold on which to evaluate the condition to have a spell.
        String with units if the method is "amount".
        Float of the quantile if the method is "quantile".
    window : int
        Number of consecutive days respecting the constraint in order to begin a spell.
        Default is 1, which is equivalent to `_bivariate_threshold_count`
    stat : {'mean', 'sum', 'max','min'}
        Statistics to apply to the remaining time dimension after resampling (e.g. Jan 1980-2010)
    stat_resample : {'mean', 'sum', 'max','min'}, optional
        Statistics to apply to the resampled input at the {group} (e.g. 1-31 Jan 1980). If `None`, the same method as `stat` will be used.
    group : {'time', 'time.season', 'time.month'}
        Grouping of the output.
        e.g. If 'time.month', the spell lengths are computed separately for each month.
    resample_before_rl : bool
        Determines if the resampling should take place before or after the run
        length encoding (or a similar algorithm) is applied to runs.

    Returns
    -------
    xr.DataArray, [units of the sampling frequency]
        {stat} of spell length distribution when the first variable is {op1} the {method1} {thresh1}
        and the second variable is {op2} the {method2} {thresh2} for {window} consecutive day(s).
```

File: xclim/sdba/properties.py [hunk 4]
```diff
@@ -730,6 +813,245 @@ def _first_output(ds, *, dim, index, corr_type):
...
+    group : {'time', 'time.season', 'time.month'}
+      Grouping of the output.
+      e.g. For 'time.month', the correlation would be calculated on each month separately,
+      but with all the years together.
+
+    Returns
+    -------
+    xr.DataArray, [dimensionless]
+      {stat} number of days when the first variable is {op1} the {method1} {thresh1}
+      and the second variable is {op2} the {method2} {thresh2} for {window} consecutive day(s).
```
Comment (Zeitsperre) [inline xclim/sdba/properties.py:1026]:
```suggestion
    da1 : xr.DataArray
        First variable on which to calculate the diagnostic.
    da2 : xr.DataArray
        Second variable on which to calculate the diagnostic.
    method1 : {'amount', 'quantile'}
        Method to choose the threshold.
        'amount': The threshold is directly the quantity in {thresh}. It needs to have the same units as {da}.
        'quantile': The threshold is calculated as the quantile {thresh} of the distribution.
    method2 : {'amount', 'quantile'}
        Method to choose the threshold.
        'amount': The threshold is directly the quantity in {thresh}. It needs to have the same units as {da}.
        'quantile': The threshold is calculated as the quantile {thresh} of the distribution.
    op1 : {">", "<", ">=", "<="}
        Operation to verify the condition for a spell.
        The condition for a spell is variable {op} threshold.
    op2 : {">", "<", ">=", "<="}
        Operation to verify the condition for a spell.
        The condition for a spell is variable {op} threshold.
    thresh1 : str or float
        Threshold on which to evaluate the condition to have a spell.
        String with units if the method is "amount".
        Float of the quantile if the method is "quantile".
    thresh2 : str or float
        Threshold on which to evaluate the condition to have a spell.
        String with units if the method is "amount".
        Float of the quantile if the method is "quantile".
    stat : {'mean', 'sum', 'max','min'}
        Statistics to apply to the remaining time dimension after resampling (e.g. Jan 1980-2010)
    stat_resample : {'mean', 'sum', 'max','min'}, optional
        Statistics to apply to the resampled input at the {group} (e.g. 1-31 Jan 1980).
        If `None`, the same method as `stat` will be used.
    group : {'time', 'time.season', 'time.month'}
        Grouping of the output.
        e.g. For 'time.month', the correlation would be calculated on each month separately,
        but with all the years together.

    Returns
    -------
    xr.DataArray, [dimensionless]
        {stat} number of days when the first variable is {op1} the {method1} {thresh1}
        and the second variable is {op2} the {method2} {thresh2} for {window} consecutive day(s).
```

**Review by Zeitsperre — APPROVED** [review]

File: xclim/sdba/properties.py [hunk 5]
```diff
@@ -386,6 +405,70 @@ def _spell_stats(ds, *, dim, method, thresh, op, freq, resample_before_rl, stat)
...
+      Str with units if the method is "amount".
+      Float of the quantile if the method is "quantile".
+    stat : {'mean', 'sum', 'max','min'}
+      Statistics to apply to the remaining time dimension after resampling (e.g. Jan 1980-2010)
+    stat_resample : {'mean', 'sum', 'max','min'}, optional
+      Statistics to apply to the resampled input at the {group} (e.g. 1-31 Jan 1980). If `None`, the same method as `stat` will be used.
+    group : {'time', 'time.season', 'time.month'}
+      Grouping of the output.
+      e.g. For 'time.month', the correlation would be calculated on each month separately,
+      but with all the years together.
```
Comment (Zeitsperre) [inline xclim/sdba/properties.py:444]: Went by this too fast, my bad!

```suggestion
    da : xr.DataArray
        Variable on which to calculate the diagnostic.
    method : {'amount', 'quantile'}
        Method to choose the threshold.
        'amount': The threshold is directly the quantity in {thresh}. It needs to have the same units as {da}.
        'quantile': The threshold is calculated as the quantile {thresh} of the distribution.
    op : {">", "<", ">=", "<="}
        Operation to verify the condition for a spell.
        The condition for a spell is variable {op} threshold.
    thresh : str or float
        Threshold on which to evaluate the condition to have a spell.
        String with units if the method is "amount".
        Float of the quantile if the method is "quantile".
    stat : {'mean', 'sum', 'max','min'}
        Statistics to apply to the remaining time dimension after resampling (e.g. Jan 1980-2010)
    stat_resample : {'mean', 'sum', 'max','min'}, optional
        Statistics to apply to the resampled input at the {group} (e.g. 1-31 Jan 1980). If `None`, the same method as `stat` will be used.
    group : {'time', 'time.season', 'time.month'}
        Grouping of the output.
        e.g. For 'time.month', the correlation would be calculated on each month separately,
        but with all the years together.
```

**Review by aulemahal — APPROVED** [review]
This code looks good.

**Pull request merged**